### PR TITLE
Fix/Update QgsServerFilter cookbook documentation

### PR DESCRIPTION
**Pull request by dmarteau**
Update QgsServerFilter documentation according to https://github.com/qgis/QGIS/pull/45158

Also fix some errors about how QgsServerFilter callbacks are called during the processing of the response.

### CODE REVIEW

**Review by DelazJ — COMMENTED** [review]
Thanks @dmarteau
I made some "cosmetic" comments. I'll let server gurus approve the contents.

File: docs/pyqgis_developer_cookbook/server.rst [hunk 1]
```diff
@@ -243,41 +248,40 @@ You could even substitute a core service completely by changing **SERVICE**
...
+This is called whenever any partial output is flushed from response buffer (i.e  to **FCGI** ``stdout`` 
+if the fcgi server is used) (and from there, to the client). 
+This occurs when huge content is streamed (like WFS GetFeature). In this case
+:meth:`onSendResponse()  <qgis.server.QgsServerFilter.onSendResponse>` may be called multiple times.
+
+Note that if the response is not streamed, then :meth:`onSendResponse() <qgis.server.QgsServerFilter.onSendResponse>` will not be called at all. 
+
+In all case, the last (or unique) chunk will be sent to client after a call to :meth:`onResponseComplete() <qgis.server.QgsServerFilter.onResponseComplete>`
+
+Returning ``False`` from will prevent flushing of data to the client. This is desirable when a plugin
```
[the remark this answers is not in the document]
Comment (dmarteau) [reply docs/pyqgis_developer_cookbook/server.rst:263]: Typo, removed "from"

File: docs/pyqgis_developer_cookbook/server.rst [hunk 2]
```diff
@@ -243,41 +248,40 @@ You could even substitute a core service completely by changing **SERVICE**
...
-request is ready to be sent to the client. As discussed above, this is normally
-called before :meth:`sendResponse() <qgis.server.QgsServerFilter.sendResponse>`
-except for streaming services (or other plugin
-filters) that might have called
-:meth:`sendResponse() <qgis.server.QgsServerFilter.sendResponse>` earlier.
+request is ready to be sent to the client.
+As discussed above, this method will be called before the last (or unique) chunk of
+data is sent to the client. 
+For streaming services, multiple calls to  :meth:`onSendResponse() <qgis.server.QgsServerFilter.onSendResponse>`
+migth have been called
```
[the remark this answers is not in the document]
Comment (dmarteau) [reply docs/pyqgis_developer_cookbook/server.rst:275]: Fixed

File: docs/pyqgis_developer_cookbook/server.rst [hunk 3]
```diff
@@ -243,41 +248,40 @@ You could even substitute a core service completely by changing **SERVICE**
...
-responseComplete
-^^^^^^^^^^^^^^^^
+This is called whenever any partial output is flushed from response buffer (i.e  to **FCGI** ``stdout`` 
+if the fcgi server is used) (and from there, to the client). 
+This occurs when huge content is streamed (like WFS GetFeature). In this case
+:meth:`onSendResponse()  <qgis.server.QgsServerFilter.onSendResponse>` may be called multiple times.
+
+Note that if the response is not streamed, then :meth:`onSendResponse() <qgis.server.QgsServerFilter.onSendResponse>` will not be called at all. 
+
+In all case, the last (or unique) chunk will be sent to client after a call to :meth:`onResponseComplete() <qgis.server.QgsServerFilter.onResponseComplete>`
```
[the remark this answers is not in the document]
Comment (dmarteau) [reply docs/pyqgis_developer_cookbook/server.rst:261]: Fixed

File: docs/pyqgis_developer_cookbook/server.rst [hunk 4]
```diff
@@ -243,41 +248,40 @@ You could even substitute a core service completely by changing **SERVICE**
...
-is typically also an option,
-:meth:`sendResponse() <qgis.server.QgsServerFilter.sendResponse>` is the only
-viable option in case of streaming services.
-
-responseComplete
-^^^^^^^^^^^^^^^^
+This is called whenever any partial output is flushed from response buffer (i.e  to **FCGI** ``stdout`` 
+if the fcgi server is used) (and from there, to the client). 
+This occurs when huge content is streamed (like WFS GetFeature). In this case
+:meth:`onSendResponse()  <qgis.server.QgsServerFilter.onSendResponse>` may be called multiple times.
```
[the remark this answers is not in the document]
Comment (dmarteau) [reply docs/pyqgis_developer_cookbook/server.rst:257]: Fixed

File: docs/pyqgis_developer_cookbook/server.rst [hunk 5]
```diff
@@ -243,41 +248,40 @@ You could even substitute a core service completely by changing **SERVICE**
...
-for direct manipulation of core service’s
-output and while :meth:`responseComplete() <qgis.server.QgsServerFilter.responseComplete>`
-is typically also an option,
-:meth:`sendResponse() <qgis.server.QgsServerFilter.sendResponse>` is the only
-viable option in case of streaming services.
-
-responseComplete
-^^^^^^^^^^^^^^^^
+This is called whenever any partial output is flushed from response buffer (i.e  to **FCGI** ``stdout`` 
+if the fcgi server is used) (and from there, to the client).
```
[the remark this answers is not in the document]
Comment (dmarteau) [reply docs/pyqgis_developer_cookbook/server.rst:255]: Removed parentheses around 'and from... ...client'.

**Comment by DelazJ** [comment]
Thanks @dmarteau
I approved to trigger the build checks.

**Comment by dmarteau** [comment]
I don't know how to fix:
```
/home/runner/work/QGIS-Documentation/QGIS-Documentation/docs/pyqgis_developer_cookbook/server.rst:369: 
WARNING: py:meth reference target not found: qgis.server.QgsServerFilter.onSendResponse
```

**Comment by DelazJ** [comment]
No worries. As soon as the code is merged in master and pyQGIS API docs generated, those references will be available

**Comment by DelazJ** [comment]
@dmarteau Now that the feature PR is merged, can you rebase this one please?
@elpaso OK to merge?

File: docs/pyqgis_developer_cookbook/server.rst [hunk 6]
```diff
@@ -197,38 +197,43 @@ From this point, you might find useful a quick look to the
...
+* :meth:`onRequestReady() <qgis.server.QgsServerFilter.onRequestReady>`
+* :meth:`onResponseComplete() <qgis.server.QgsServerFilter.onResponseComplete>`
+* :meth:`onSendResponse() <qgis.server.QgsServerFilter.onSendResponse>`
 
 All filters have access to the request/response object
 (:class:`QgsRequestHandler <qgis.server.QgsRequestHandler>`)
 and can manipulate all its properties (input/output) and
 raise exceptions (while in a quite particular way as we’ll see below).
 
+All these methods return a boolean value indicating if the call should be propagated the subsequent
```
Comment (DelazJ) [inline docs/pyqgis_developer_cookbook/server.rst:209]:
```suggestion
All these methods return a boolean value indicating if the call should be propagated to the subsequent
```

File: docs/pyqgis_developer_cookbook/server.rst [hunk 7]
```diff
@@ -243,41 +248,43 @@ You could even substitute a core service completely by changing **SERVICE**
...
 This is called once when core services (if hit) finish their process and the
-request is ready to be sent to the client. As discussed above, this is normally
-called before :meth:`sendResponse() <qgis.server.QgsServerFilter.sendResponse>`
-except for streaming services (or other plugin
-filters) that might have called
-:meth:`sendResponse() <qgis.server.QgsServerFilter.sendResponse>` earlier.
+request is ready to be sent to the client.
+As discussed above, this method will be called before the last (or unique) chunk of
+data is sent to the client. 
+For streaming services, multiple calls to :meth:`onSendResponse() <qgis.server.QgsServerFilter.onSendResponse>` migth have been called.
```
Comment (DelazJ) [inline docs/pyqgis_developer_cookbook/server.rst:276]:
```suggestion
For streaming services, multiple calls to :meth:`onSendResponse() <qgis.server.QgsServerFilter.onSendResponse>` might have been called.
```

**Review by DelazJ — COMMENTED** [review]
Typos

**Pull request merged**